From e1d97756296ac12e0c4495890250fae34069da91 Mon Sep 17 00:00:00 2001
From: Zhang Xiao <xiao.zhang@windriver.com>
Date: Wed, 30 Dec 2009 11:04:12 +0800
Subject: [PATCH] Fix bugs of mxc nand flash driver

If only spare data was considered and no error occurred on it, no more error
correct operations needed. Modify the error byte location logic and make the
col value more safe.

Signed-off-by: Zhang Xiao <xiao.zhang@windriver.com>
---
 drivers/mtd/nand/mxc_nand.c |    7 +++++--
 1 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/mtd/nand/mxc_nand.c b/drivers/mtd/nand/mxc_nand.c
index 468557d..5701c2e 100644
--- a/drivers/mtd/nand/mxc_nand.c
+++ b/drivers/mtd/nand/mxc_nand.c
@@ -305,9 +305,9 @@ static void mxc_nd_correct_error(struct mxc_nand_host *host,
 	/* Set the pointer for main / spare area */
 	if (!bSpareOnly)
 		buf = host->regs + MAIN_AREA0 +
-			col + (256 * buf_id);
+			(col & 0xff) * 2 + (512 * buf_id);
 	else
-		buf = host->regs + SPARE_AREA0 + col + (8 * buf_id);
+		buf = host->regs + SPARE_AREA0 + (col & 0x1) * 2 + (16 * buf_id);
 
 	/* Fix the data */
 	tmp = readw(buf);
@@ -332,6 +332,9 @@ static void mxc_nd_correct_ecc(struct mxc_nand_host *host,
 	DEBUG(MTD_DEBUG_LEVEL3,
 			"mxc_nd_correct_ecc (Ecc status=%x)\n", ecc_status);
 
+	if(spare && !(ecc_status & 0x3))
+		return;
+
 	if (((ecc_status & 0xC) == MAIN_SINGLEBIT_ERROR)
 			|| ((ecc_status & 0x3) == SPARE_SINGLEBIT_ERROR)) {
 		if (host->ecc_disabled) {
-- 
1.6.5.2

